### DISCUSSION TITLE
Trouble with perl and cgi

### DISCUSSION THREAD

**Discussion by new2brew** [Homebrew/discussions#5758] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: cc0b5d6d05d1f3af5b2f53e9fbb1e7bfa0c332ab
Last commit: 5 days ago
Core tap JSON: 22 Nov 06:34 UTC
Core cask tap JSON: 22 Nov 05:39 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 3.3.6 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: quad-core 64-bit icelake
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.1.1-x86_64
CLT: 16.1.0.0.1.1729049160
Xcode: N/A
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I am running homebrew with html and cgi files that I have imported from a Lubuntu setup where they work fine. I have an initial html file that when submitted  calls a cgi file that is also a perl file.... here are the first few lines:

#!/usr/local/Cellar/perl -w

#------------------------------
#  interactive.pl
#------------------------------

print "Content-Type: text/html \n\n";

use File::Copy;
use warnings;
no warnings ('uninitialized', 'substr');
use strict;
use Data::Dumper;
use File::Find;
use CGI;
use CGI qw(:cgi-lib :standard);
use CGI::Carp qw(fatalsToBrowser)

I am not an experienced homebrew user but I think I understand some of the fundamentals. When this cgi file gets called it just displays the raw code, not in html format. I believe I had similar issues in Lubuntu but was able to fix the problems by adding lines to enable cgi in the httpd.conf file and use sudo a2enmode cgid. The a2enmode command is not avaialble now in homebrew. Disclaimer - I have not revisted this code in years and at the time it was not a logical path to success to get the code working in Lubuntu... it was a bit wacamole. However I'm nothing but persistent and all I need to start is to know how I need to enable the cgi and perl aspects of homebrew... I have done brew install perl and changed lines in httpd.conf to try to get this to work .... I have done brew link perl... I have changed the bash line at the beginning of the cgi file (#!/usr/local/Cellar/perl -w) over and over.... I am at a loss. I have loaded and unloaded modules in httpd.conf - nothing seems to work. If anyone could give me some pointers I'd appreciate it..... As a 72 year old wondering why I am doing this sometimes... so cut me some slack... I think if I can get this cgi perl problem fixed it will solve all the issues.

**Comment by gromgit** [comment]
> ```
> #!/usr/local/Cellar/perl -w
> ```

That path points to a directory, not a perl binary. Since you have Homebrew `perl` installed, this should also certainly be:
```
#!/usr/local/bin/perl -w
```

**Reply by new2brew** [reply]
Thanks for the reply, but I had tried that earlier with no luck and I again retried it but no luck.... maybe I have made changes to httpd.conf to compensate and they could have screwed things up. i will go back and recheck that. Anymore suggestions would be appreciated

**Reply by new2brew** [reply]
I ran the file from the command line at got the following: ---------
Can't locate CGI.pm in @INC (you may need to install the CGI module) (@INC entries checked: /usr/local/opt/perl/lib/perl5/site_perl/5.40/darwin-thread-multi-2level /usr/local/opt/perl/lib/perl5/site_perl/5.40 /usr/local/opt/perl/lib/perl5/5.40/darwin-thread-multi-2level /usr/local/opt/perl/lib/perl5/5.40 /usr/local/lib/perl5/site_perl/5.40/darwin-thread-multi-2level /usr/local/lib/perl5/site_perl/5.40) at ./interactive.cgi line 17.
BEGIN failed--compilation aborted at ./interactive.cgi line 17.------------
Line 17 is simply use:CGI

**Reply by gromgit** [reply]
Looks like you don't have the CGI Perl module installed. The paths listed in the error message also indicate that you didn't do the steps listed in the ***Caveats*** section of `brew info perl`. You should probably do those first, but since you've asked to be cut some slack, I'll note that you'll need to do two things with the last command in the caveats (i.e. `eval ...`):

1. Add that command to your login shell's startup file (`~/.profile` if bash, `~/.zshrc` if zsh) to ensure _future_ shell sessions are set up to use `local::lib`.
2. Run that same command manually, to ensure your _current_ shell session is set up to use `local::lib`.

Once all that is done, then `cpan install CGI` should install the necessary module with a TON of output but no errors. You can confirm that by running `perl -e 'use CGI'`; it should print nothing if all is well.

**Comment by new2brew** [comment]
Wow. Thanks ..  I’m not totally done but I feel like almost there. I
noticed those caveats but I didn’t pay too much attention to them. Then
they were gone. What a huge install that was of cpan perl. And I got no
errors. I would never have had a clue that was how to install Perl. I
thought the brew install was the solution. I think I have some issues still
with the zshrc file. I have 2 lines in that file. The one you mentioned and
another. I can’t seem to figure out the minimum basic config of that file.
I get so many suggestions in google but not just a basic homebrew config.
Thanks again.

**Reply by new2brew** [reply]
Finally Success... --- your help was very much appreciated. I kept running into the same issue BUT I new CGI was working (through your advice). So I concentrated on the httpd.conf file and finally had success. Unfortunately I don't know if I can pin it down to one thing in the conf file but I believe it was the two </IfModule>
<IfModule mpm_prefork_module> lines. Anyway I can now breathe a little easier... Thanks
